src/PostgREST/Auth.hs

```diff
@@ -107,16 +106,52 @@ middleware appState app req respond = do
   let token  = fromMaybe "" $ Wai.extractBearerAuth =<< lookup HTTP.hAuthorization (Wai.requestHeaders req)
       parseJwt = runExceptT $ parseToken conf (LBS.fromStrict token) time >>= parseClaims conf
 
-  if configDbPlanEnabled conf
-    then do
-      (dur,authResult) <- timeItT parseJwt
-      let req' = req { Wai.vault = Wai.vault req & Vault.insert authResultKey authResult & Vault.insert jwtDurKey dur }
-      app req' respond
-    else do
-      authResult <- parseJwt
-      let req' = req { Wai.vault = Wai.vault req & Vault.insert authResultKey authResult }
-      app req' respond
-
+-- If DbPlanEnabled -> calculate JWT validation time
+-- If JwtCaching    -> cache JWT validation result
+  req' <- case (configDbPlanEnabled conf, configJwtCacheMaxLifetime conf) of
+    (True, 0)   -> do
+          (dur, authResult) <- timeItT parseJwt
+          return $ req { Wai.vault = Wai.vault req & Vault.insert authResultKey authResult & Vault.insert jwtDurKey dur }
+
+    (True, maxLifetime)  -> do
+          (dur, authResult) <- timeItT $ getJWTFromCache appState token maxLifetime parseJwt
+          return $ req { Wai.vault = Wai.vault req & Vault.insert authResultKey authResult & Vault.insert jwtDurKey dur }
+
+    (False, 0)  -> do
+          authResult <- parseJwt
+          return $ req { Wai.vault = Wai.vault req & Vault.insert authResultKey authResult }
+
+    (False, maxLifetime) -> do
+          authResult <- getJWTFromCache appState token maxLifetime parseJwt
+          return $ req { Wai.vault = Wai.vault req & Vault.insert authResultKey authResult }
+
+  app req' respond
+
+-- Used to extract JWT exp claim and add to JWT Cache
+getTimeSpec :: AuthResult -> Int -> Maybe CL.TimeSpec
+getTimeSpec res maxLifetime = do
+  expireJSON <- KM.lookup "exp" (authClaims res)
+  case expireJSON of
+    JSON.Number seconds -> Just $ unsafePerformIO (posixToCurrentTime seconds)
+    _                   -> Just $ CL.TimeSpec (fromIntegral maxLifetime :: Int64) 0 -- set timeSpec to 0 so it expires immediately, hence not cached
```

Comment:
Hm, this part is not clear to me. So as we discussed before, if there's no `exp`claim, the cache lifetime should be `maxLifetime` and not 0 right?

Reply:
Adding a test for ensuring caching works for a JWT without exp would also be good.

Reply:
Oh, my bad. It is actually `maxLifetime`. I forgot to remove the comment after changing that.